File: kotlin/algorithms/src/main/kotlin/searching/BinarySearch.kt
package searching

import kotlin.math.floor

fun search(haystack: IntArray, needle: Int): Boolean {
  var lo = 0;
  var hi = haystack.size;

  do {
    val m = floor((lo + (hi - lo) / 2).toDouble());
    val v = haystack[m.toInt()];

    if (v == needle) {
      return true;
    } else if (v > needle) {
      hi = m.toInt();
    } else {
      lo = (m + 1).toInt();
    }

  } while (lo < hi);

  return false;
}

File: dart/lib/binary_search.dart
bool bsList(List<int> haystack, int needle) {
  int lo = 0;
  int hi = haystack.length;

  do {
    final m = (lo + (hi - lo) / 2).floor();
    final v = haystack[m];

    if (v == needle) {
      return true;
    } else if (v > needle) {
      hi = m;
    } else {
      lo = m + 1;
    }
  } while (lo < hi);

  return false;
}
